src/pool-cl/libraries/CLPoolGetters.sol

```diff
@@ -3,9 +3,40 @@
 pragma solidity ^0.8.24;
 
 import {CLPool} from "./CLPool.sol";
+import {CLPosition} from "./CLPosition.sol";
 import {Tick} from "./Tick.sol";
 
 library CLPoolGetters {
+    struct PositionInfo {
+        address owner;
+        int24 tickLower;
+        int24 tickUpper;
+        uint128 liquidity;
+        uint256 feeGrowthInside0LastX128;
+        uint256 feeGrowthInside1LastX128;
+    }
+
+    function getPoolNumPositions(CLPool.State storage pool) internal view returns (uint256) {
+        return pool.positionKeys.length;
+    }
+
+    function getPoolPositionInfo(CLPool.State storage pool, uint256 index)
+        internal
+        view
+        returns (PositionInfo memory)
+    {
+        CLPool.PositionKey memory key = pool.positionKeys[index];
+        CLPosition.Info memory info = pool.positions[CLPosition.hashKey(key.owner, key.tickLower, key.tickUpper)];
+        return PositionInfo({
+            owner: key.owner,
+            tickLower: key.tickLower,
+            tickUpper: key.tickUpper,
+            liquidity: info.liquidity,
+            feeGrowthInside0LastX128: info.feeGrowthInside0LastX128,
+            feeGrowthInside1LastX128: info.feeGrowthInside1LastX128
+        });
+    }
+
```

Comment:
So what u want here is basically an array to keep track of all non-empty position, so that you can:

```
for i = 0 to length - 1:
    gather and calc farming info for each position 
```

But why not just listen the event `ModifyLiquidity` in a off-chain manner, and commit update to db instead

```
event ModifyLiquidity(
      PoolId indexed id,
      address indexed sender,
      int24 tickLower,
      int24 tickUpper,
      int256 liquidityDelta,
      int128 feesOwed0,
      int128 feesOwed1
  );
```